app/cni/cmd/kuma-cni/main.go

```diff
@@ -0,0 +1,266 @@
+		"pod", string(k8sArgs.K8S_POD_NAME),
+		"namespace", string(k8sArgs.K8S_POD_NAMESPACE),
+		"annotations", annotations)
+}
+
+func isSidecarInjectedAnnotationAbsent(annotations map[string]string) bool {
+	excludePod := false
+	val, ok := annotations["kuma.io/sidecar-injected"]
+	if !ok || val != "true" {
+		log.V(1).Info("pod excluded due to lack of 'kuma.io/sidecar-injected: true' annotation")
+		excludePod = true
+	}
+	return excludePod
+}
+
+func isInitContainerPresent(initContainersMap map[string]struct{}, k8sArgs K8sArgs) bool {
+	excludePod := false
+	// Check if kuma-init container is present; in that case exclude pod
+	if _, present := initContainersMap["kuma-init"]; present {
+		log.V(1).Info("pod excluded due to being already injected with kuma-init container",
+			"pod", string(k8sArgs.K8S_POD_NAME),
+			"namespace", string(k8sArgs.K8S_POD_NAMESPACE))
+		excludePod = true
+	}
+	return excludePod
+}
+
+func getPodInfoWithRetries(ctx context.Context, conf *PluginConf, k8sArgs K8sArgs) (int, map[string]struct{}, map[string]string, error) {
+	client, err := newKubeClient(*conf)
+	if err != nil {
+		return 0, nil, nil, err
+	}
+	log.V(1).Info("created Kubernetes client", "client", client)
+
+	var containerCount int
+	var initContainersMap map[string]struct{}
+	var annotations map[string]string
+	var k8sErr error
+	for attempt := 1; attempt <= podRetrievalMaxRetries; attempt++ {
+		containerCount, initContainersMap, annotations, k8sErr = getKubePodInfo(ctx, client, string(k8sArgs.K8S_POD_NAME), string(k8sArgs.K8S_POD_NAMESPACE))
+		log.V(1).Info("container count in a pod", "count", containerCount)
+		if k8sErr == nil {
+			break
+		}
+		log.Error(k8sErr, "waiting for pod metadata", "attempt", attempt)
+		time.Sleep(podRetrievalInterval)
+	}
+	if k8sErr != nil {
+		log.Error(k8sErr, "failed to get pod data")
+		return 0, nil, nil, k8sErr
+	}
+	return containerCount, initContainersMap, annotations, nil
+}
+
+func shouldExcludePod(excludedNamespaces []string, podNamespace types.UnmarshallableString) bool {
+	excludePod := false
+	for _, excludeNs := range excludedNamespaces {
+		if string(podNamespace) == excludeNs {
+			excludePod = true
+			break
+		}
+	}
+	return excludePod
+}
+
+func logContainerInfo(args string, containerID string, k8sArgs K8sArgs) {
+	log.Info("getting identifiers with arguments: %s", "arguments", args)
+	log.Info("loaded k8s arguments: %v", "k8s args", k8sArgs)
+	log.Info("container information",
+		"containerID", containerID,
+		"pod", string(k8sArgs.K8S_POD_NAME),
+		"namespace", string(k8sArgs.K8S_POD_NAMESPACE))
+}
+
+func logPrevResult(conf *PluginConf) {
+	var loggedPrevResult interface{}
+	if conf.PrevResult == nil {
+		loggedPrevResult = "none"
+	} else {
+		loggedPrevResult = conf.PrevResult
+	}
+
+	log.V(1).Info("cmdAdd config parsed",
+		"version", conf.CNIVersion,
+		"prevResult", loggedPrevResult)
+}
+
+func cmdGet(args *skel.CmdArgs) error {
+	log.Info("cmdGet not implemented")
+	// TODO: implement
```

Comment:
is there an issue to implement it?

Reply:
TBH other CNIs don't implement it and we've lived without it for so long so I'm removing that todo. I'll create an issue and we can talk about this during triage.

Reply:
Issue created here: https://github.com/kumahq/kuma/issues/4606